### PR TITLE
Changes pulling out django_tenant_schemas plugin

### PR DESCRIPTION
**Pull request by wcmitchell**
Unit tests passing locally, manual testing exercises yet to be done, but wanted to get a working push up and under the other devs' eyes. It's a big mess, touches nearly everything, but hopefully working!

Signed-off-by: Chris Mitchell <cmitchel@redhat.com>

## Link(s) to Jira
- https://issues.redhat.com/browse/RHCLOUD-18022
- https://issues.redhat.com/browse/RHCLOUD-18026

## Description of Intent of Change(s)
Now that we've collapsed everything into the public schema, we should be able to clean up the code and remove references to django_tenant_schemas

## Local Testing
How can the feature be exercised? 
How can the bug be exploited and fix confirmed?
Is any special local setup required?

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ x ] are there any pre/post merge actions required? if so, document here.
  - Two new migrations to rename the 'schema_name' field to 'tenant_name'
- [ x ] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ x ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  -  if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: rbac/management/utils.py [hunk 1]
```diff
@@ -113,7 +114,7 @@ def groups_for_principal(principal, tenant, **kwargs):
         return set()
     assigned_group_set = principal.group.all()
     if settings.SERVE_FROM_PUBLIC_SCHEMA:
```
Comment (astrozzc) [inline rbac/management/utils.py:116]: Maybe we could remove this now

**Review by astrozzc — COMMENTED** [review]
Nice 👍 I have have few questions

**Comment by coderbydesign** [comment]
@wcmitchell should we remove all the conditional `if settings.SERVE_FROM_PUBLIC_SCHEMA:` checks, and the setting now, or do we have that logged somewhere else?

**Comment by coderbydesign** [comment]
@wcmitchell should we remove all the conditional `if settings.SERVE_FROM_PUBLIC_SCHEMA:` checks, and the setting now, or do we have that logged somewhere else?

File: rbac/internal/views.py [hunk 2]
```diff
@@ -51,23 +50,22 @@ def destructive_ok():
...
     # the search_path on the query will fall back to using the public schema, in
     # which case there will be custom groups/roles, and we won't be able to propertly
     # prune the tenant which has been created without a valid schema
-    if not schema_exists(schema_name):
-        return False
+    tenant = Tenant.objects.get(tenant_name=tenant_name)
 
-    return (Role.objects.filter(system=True).count() != Role.objects.count()) or (
-        Group.objects.filter(system=True).count() != Group.objects.count()
+    return (Role.objects.filter(system=False, tenant=tenant).count() != 0) or (
```
Comment (coderbydesign) [inline rbac/internal/views.py:61]: Nice catch on the logic change with the role/group counts now against public.

File: Makefile [hunk 3]
```diff
@@ -99,7 +99,7 @@ make-migrations:
 	DJANGO_READ_DOT_ENV_FILE=True $(PYTHON) $(PYDIR)/manage.py makemigrations api management
 
 run-migrations:
-	DJANGO_READ_DOT_ENV_FILE=True $(PYTHON) $(PYDIR)/manage.py migrate_schemas --executor=parallel
+	DJANGO_READ_DOT_ENV_FILE=True $(PYTHON) $(PYDIR)/manage.py migrate
```
[the remark this answers is not in the document]
Comment (wcmitchell) [reply Makefile:102]: Right! The migrate-schemas command was provided by DTS, and pulling it out we need to revert to the base Django migrate command.

File: rbac/api/migrations/0011_rename_schema_name_to_tenant_name.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("api", "0010_explicitly_define_tenant_schema_name"),
+    ]
+
+    operations = [
+        migrations.RenameField(model_name="tenant", old_name="schema_name", new_name="tenant_name",),
```
[the remark this answers is not in the document]
Comment (wcmitchell) [reply rbac/api/migrations/0011_rename_schema_name_to_tenant_name.py:13]: I think so, yeah. I'd originally tried doing it all in one migration, but got complaints during the migration run. I can try rolling back to one migration and log the issues if we need. If I remember right, it was having to do with adding a new field (even though it was a rename migration) without allowing null data? Whatever it was, it lead me to that same stackoverflow question.

File: rbac/internal/views.py [hunk 5]
```diff
@@ -51,23 +50,22 @@ def destructive_ok():
...
-def tenant_is_modified(schema_name):
+def tenant_is_modified(tenant_name):
     """Determine whether or not the tenant is modified."""
     # we need to check if the schema exists because if we don't, and it doesn't exist,
     # the search_path on the query will fall back to using the public schema, in
     # which case there will be custom groups/roles, and we won't be able to propertly
     # prune the tenant which has been created without a valid schema
-    if not schema_exists(schema_name):
-        return False
+    tenant = Tenant.objects.get(tenant_name=tenant_name)
```
[the remark this answers is not in the document]
Comment (wcmitchell) [reply rbac/internal/views.py:59]: Good thought, I'll add that in!

File: rbac/api/migrations/0011_rename_schema_name_to_tenant_name.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (coderbydesign) [reply rbac/api/migrations/0011_rename_schema_name_to_tenant_name.py:13]: Got it. Yeah, that makes sense. It sounds like Django recognizes it's a no-op and doesn't actually create a new column.

**Comment by wcmitchell** [comment]
> @wcmitchell a few questions so far:
> 
>     * should we remove all the conditional `if settings.SERVE_FROM_PUBLIC_SCHEMA:` checks, and the setting now, or do we have that logged somewhere else?
> 
>     * can we remove the `django-tenant-schemas` library from the Pipfile?
> 
>     * there are a few instances of still using `with tenant_context` in `/commands/*` - if we're pulling DTS out completely, we'll no longer have that import (or need for it). we should probably just clean those commands up/remove them if they are schema-specific? based on a project search, it looks like those are the only other references of imports of `from tenant_schemas`
> 
>     * I found a few other references to `schema_name`, some are just cosmetic, but a few are actual column references:
>       
>       * https://github.com/wcmitchell/insights-rbac/blob/remove_django_tenant_schemas/rbac/management/management/commands/sync_schemas.py#L206
>       * https://github.com/wcmitchell/insights-rbac/blob/remove_django_tenant_schemas/rbac/management/management/commands/init_tenant.py#L38-L45
>       * https://github.com/wcmitchell/insights-rbac/blob/remove_django_tenant_schemas/rbac/management/tasks.py#L61-L63
>       * https://github.com/wcmitchell/insights-rbac/blob/remove_django_tenant_schemas/rbac/internal/views.py#L148-L157

Went through and fixed most of this up. Found we can't remove django-tenant-schemas from the pipfile yet because our migrations reference it. When/if we do a migration squash that should resolve that.

As to the SERVE_FROM_PUBLIC_SCHEMA toggle, I had been planning on pulling that out in a future PR, trying to keep these changes as narrowly scoped as I could. Don't have anything logged for that work yet, but can definitely get something created.

**Comment by astrozzc** [comment]
LGTM, looks like there are some lint issues left

File: rbac/rbac/settings.py [hunk 6]
```diff
@@ -166,8 +162,6 @@
 
 DATABASES = {"default": database.config()}
 
-DATABASE_ROUTERS = ("tenant_schemas.routers.TenantSyncRouter",)
-
 #
 TENANT_MODEL = "api.Tenant"
```
Comment (coderbydesign) [inline rbac/rbac/settings.py:166]: Can we pull this out?

File: rbac/rbac/settings.py [hunk 7]
```diff
@@ -120,8 +118,6 @@
 
 TENANT_APPS = ("management",)
```
Comment (coderbydesign) [inline rbac/rbac/settings.py:119]: Can we pull this out?

File: rbac/rbac/settings.py (hunk 6 above)
Comment (wcmitchell) [reply rbac/rbac/settings.py:166]: Yep! Will give it a shot

File: rbac/rbac/settings.py (hunk 7 above)
Comment (wcmitchell) [reply rbac/rbac/settings.py:119]: Same as above

**Comment by coderbydesign** [comment]
@wcmitchell it looks like we need to update the init container commands: https://github.com/RedHatInsights/insights-rbac/blob/master/deploy/init-container-setup.sh

File: rbac/management/seeds.py [hunk 8]
```diff
@@ -61,20 +56,10 @@ def run_seeds(seed_type, schema_list=None):
...
-                )
-                future = executor.submit(seed_functions[seed_type], tenant)
-                completed_log_message = (
-                    f"Finished seeding {seed_type} changes for tenant "
-                    f"{tenant.schema_name} [{idx + 1} of {tenant_count}]."
-                )
-                future.add_done_callback(partial(on_complete, completed_log_message, tenant))
+
+            logger.info(f"Seeding {seed_type} changes.")
+            future = executor.submit(seed_functions[seed_type])
```
Comment (astrozzc) [inline rbac/management/seeds.py:61]: Here the function seed_roles, seed_group, seed_permissions still require tenant as input. Which will not run.

But this is included in my PR, should we merge this and get it fixed in my PR, or fix here? Both works for me. @wcmitchell @coderbydesign

**Review by astrozzc — APPROVED** [review]
LGTM

**Comment by wcmitchell** [comment]
retest

**Pull request merged**